draft-irtf-cfrg-hybrid-kems.md

```diff
@@ -537,175 +545,326 @@ and L arguments.
 * `Label` - A byte string used to label the specific combination of the above
-  constituents being used.  This value should be registered in the Hybrid KEM
+  constituents being used, as well as which framework is being instantiated.
+  This value should be registered in the Hybrid KEM
   Labels IANA registry to avoid conflict with other instantiations (see
   {{iana-considerations}}.
 
-The KEMs, groups, KDFs, and PRGs MUST meet the security requirements listed in
-{{hybrid-ind-cca}}.
+`KEM_PQ`, `Group_T`, `PRG`, and `KDF` MUST meet the interfaces
+described in {{cryptographic-deps}} and MUST meet the security requirements
+described in {{hybrid-ind-cca}}.
 
 The constants for public values are derived from the concatenation of
 encapsulation keys and ciphertexts:
 
 ~~~
-Nek = KEM_PQ.Nek + KEM_T.Nek
-Nct = KEM_PQ.Nct + KEM_T.Nct
+KEM_H.Nek = KEM_PQ.Nek + (KEM_T.Nek or Group_T.Nelem)
+KEM_H.Nct = KEM_PQ.Nct + (KEM_T.Nct or Group_T.Nelem)
 ~~~
 
-The `Nseed` and `Nss` constants should reflect the overall security level of the
-combined KEM, with the following recommended values:
+The `Nseed` and `Nss` constants should reflect the overall security level of
+the combined KEM, with the following recommended values:
 
 ~~~
-Nseed = max(KEM_PQ.Nseed, KEM_T.Nseed)
-Nss = min(KEM_PQ.Nss, KEM_T.Nss)
+KEM_H.Nseed = max(KEM_PQ.Nseed, (KEM_T.Nseed or Group_T.Nseed))
+KEM_H.Nss = min(KEM_PQ.Nss, (KEM_T.Nss or Group_T.Nss))
 ~~~
 
-Since we use the seed as the decapsulation key, `Ndk = Nseed`.
+Since we use the seed as the decapsulation key, `Ndk = Nseed`.  For legacy cases
+where it is not possible to derive per-component decapsulation keys from a
+common seed, see {{key-generation}}.
+
+## Subroutines
+
+The four hybrid KEM framework share a substantial amount of structure, which we
+capture in a set of subroutines.
 
-Given these constituent parts, the GHP hybrid KEM is defined as follows:
+### Using a Nominal Group
+
+Hybrid KEM frameworks that use a KEM for the traditional component invoke the
+DeriveKeyPair, Encap, and Decap functions of PQ KEMs, alongside analogous
+functions of the nominal group.  The "encapsulation key" is the receiver's
+public group element; the "ciphertext" is an ephemeral group element; and
+the shared secret is the secret value resulting from an ephemeral-static
+Diffie-Hellman exchange.
+
+~~~ pseudocode
+def expandDecapsKeyG(seed):
+    seed_full = PRG(seed)
+    (seed_PQ, seed_T) = split(KEM_PQ.Nseed, Group_T.Nseed, seed_full)
+
+    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
+    dk_T = Group_T.RandomScalar(seed_T)
+    ek_T = Group_T.Exp(Group_T.g, dk_T)
 
+    return (ek_PQ, ek_T, dk_PQ, dk_T)
+
+def prepareEncapG(ek_PQ, ek_T):
+    (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)
+    sk_E = Group_T.RandomScalar(random(Group_T.Nseed))
+    ct_T = Group_T.Exp(Group_T.g, sk_E)
+    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ek_T, sk_E))
+    return (ss_PQ, ss_T, ct_PQ, ct_T)
+
+def prepareDecapG(ct_PQ, ct_T, dk_PQ, dk_T):
```

Comment:
Not sure about 'prepare', this is doing a lot of the work of `Encaps`/`Decaps`, maybe 'process'?

Reply:
"Prepare" in the sense of "preparing the inputs to the KDF"